Compatibility Mode: add vertex storage buffer and texture limits

SenorBlanco opened the pull request:
Compatibility proposal: add separate limits for the minimum number of storage buffers and storage textures in the vertex stage.

Allow them to be zero, to accommodate OpenGL devices in the field with such a limit.

Issue: https://github.com/gpuweb/gpuweb/issues/4721

kainino0x commented on proposals/compatibility-mode.md:
```diff
@@ -197,6 +197,18 @@ generate a validation error.
+
+If the number of shader variables of type `texture_storage_1d`, `texture_storage_2d`, `texture_storage_2d_array` and `texture_storage_3d` in a vertex shader exceeds the `maxVertexStageStorageTextures` limit, a validation error will occur at pipeline creation time.
+
+In Compatibility mode, these new limits will have a default of zero. In Core mode, they will default to the same values as the existing `maxStorageBuffersPerShaderStage` and `maxStorageTexturesPerShaderStage`, respectively.
```
Ideas for how exactly this works mechanically in the spec:
- ~in core, just have the two new limits default to the same defaults as the old ones, while still enforcing them.~ This does not work because it would introduce new limits that need to be raised for core content (it's a breaking change).
- in core, skip validating the two new limits. This is technically fine but very unintuitive - you'd have some limits that you think you need to set, but don't.
- in requestDevice, "reify" the limits request by upgrading, e.g. `maxVertexStageStorageBuffers` so it's always at least as high as `maxStorageBuffersPerShaderStage`
  - (I know I've suggested similar things in the past for other pairs of limits, to make devices always match the [Adapter Capability Guarantees](https://gpuweb.github.io/gpuweb/#adapter-capability-guarantees) just to polish some corners, but I can't find the past discussion for the life of me. OTOH in this proposal it would actually be very important, to avoid being a breaking change.)
- in core, just have these two new limits default to infinity / a very large value, to signal that the limits don't exist. I prefer this one as I think it reduces surprises.

teoxoy replied:
I think I'd also prefer to have them default to `u32::MAX` in core with a note saying that the per stage limits still apply.

kainino0x commented on proposals/compatibility-mode.md:
```diff
@@ -197,6 +197,18 @@ generate a validation error.
+
+In Compatibility mode, these new limits will have a default of zero. In Core mode, they will default to the same values as the existing `maxStorageBuffersPerShaderStage` and `maxStorageTexturesPerShaderStage`, respectively.
+
+In Compatibility mode, the existing `maxStorageBuffersPerShaderStage` and `maxStorageTexturesPerShaderStage` limit will apply to the Compute stage only.
```
Alternative: still apply these limits to all stages, so both limits are enforced: the effective limit is `min(maxStorageBuffersPerShaderStage, maxVertexStageStorageBuffers)`.

I somewhat prefer this because (1) then the limit does what it says, and (2) it reduces the amount of logic that changes when compat is active.

teoxoy replied:
I was proposing to add limits to the fragment stage as well so that we can increase the default limit for `maxStorageBuffersPerShaderStage` and `maxStorageTexturesPerShaderStage` in compat to 8 (https://github.com/gpuweb/gpuweb/issues/4721#issuecomment-2304348473 & https://github.com/gpuweb/gpuweb/issues/4721#issuecomment-2318325427) but I don't think there is consensus to do that yet.

kainino0x commented on proposals/compatibility-mode.md:
```diff
@@ -197,6 +197,18 @@ generate a validation error.
 other APIs only support the first vertex so only `@interpolation(flat, either)` is supported in
 compatibility mode.
 
+## 18. Introduce new `maxVertexStageStorageBuffers` and `maxVertexStageStorageTextures` limits.
```
bikeshedding:
I agree `maxStorageBuffersPerVertexStage` wouldn't make sense since there is only one vertex stage, but as I've been talking about this I find it a little confusing that the word order differs from the old limits. Maybe `maxStorageBuffersInVertexStage`?

Also as I've been typing these comments I keep mistyping as `VertexBuffer` because the words are in that order. That's not really important but just noting.

teoxoy replied:
`maxStorageBuffersInVertexStage` sounds better to me as well.

kainino0x commented:
(LGTM other than fine details)

SenorBlanco commented:
OK, I've made the requested changes.

- renamed the limits to `maxStorage<thingy>InVertexStage`.
- changed the maximum in core to be GPUSize32::Max
- clarified that the old limits still apply in all stages, and gave an example of the effective limit in the vertex stage

Along the way, this removed the comment about the old limits being Compute-only in Compat.
